app/src/test/scala/org/alephium/explorer/util/InputAddressUtilSpec.scala

```diff
@@ -0,0 +1,131 @@
+      forAll(unlockScriptProtocolP2SHGen, outputRefProtocolGen) {
+        case (p2sh, outputRef) =>
+          val assetInput = buildAssetInput(outputRef, p2sh)
+
+          val result = InputAddressUtil
+            .addressFromProtocolInput(assetInput)
+
+          result.nonEmpty is true
+
+          val lockup = protocol.vm.LockupScript.p2sh(protocol.Hash.hash(p2sh.script))
+          result.get.toString is protocol.model.Address.from(lockup).toBase58
+      }
+    }
+
+    "fail to convert P2MPKH input into an address" in new Fixture {
+      forAll(unlockScriptProtocolP2MPKHGen, outputRefProtocolGen) {
+        case (p2mpkhUnlock, outputRef) =>
+          val assetInput = buildAssetInput(outputRef, p2mpkhUnlock)
+
+          InputAddressUtil
+            .addressFromProtocolInput(assetInput) is None
+      }
+    }
+  }
+
+  "addressFromProtocolInputs" should {
+    "return None if list is empty" in new Fixture {
+      InputAddressUtil.addressFromProtocolInputs(ArraySeq.empty) is None
+    }
+
+    "return a single address if all inputs have the same p2pkh address" in new Fixture {
+      forAll(Gen.nonEmptyListOf(outputRefProtocolGen)) { outputRefs =>
+        val p2pkh       = unlockScriptProtocolP2PKHGen.sample.get
+        val assetInputs = outputRefs.map(buildAssetInput(_, p2pkh))
+
+        val result = InputAddressUtil.addressFromProtocolInputs(assetInputs)
+
+        result.get.toString is protocol.model.Address.p2pkh(p2pkh.publicKey).toBase58
+      }
+    }
+
+    "return None if > 1 address" in new Fixture {
+      forAll(Gen.nonEmptyListOf(outputRefProtocolGen), outputRefProtocolGen) {
+        case (outputRefs, outputRef) =>
+          val p2pkh       = unlockScriptProtocolP2PKHGen.sample.get
+          val assetInputs = outputRefs.map(buildAssetInput(_, p2pkh))
+
+          val otherP2pkh      = unlockScriptProtocolP2PKHGen.sample.get
+          val otherAssetInput = buildAssetInput(outputRef, otherP2pkh)
+
+          val result = InputAddressUtil.addressFromProtocolInputs(otherAssetInput +: assetInputs)
+
+          result is None
+      }
+    }
+
+    "return None if one of the address isn't extractable" in new Fixture {
+      forAll(Gen.nonEmptyListOf(outputRefProtocolGen), outputRefProtocolGen) {
+        case (outputRefs, outputRef) =>
+          val p2pkh       = unlockScriptProtocolP2PKHGen.sample.get
+          val assetInputs = outputRefs.map(buildAssetInput(_, p2pkh))
+
+          val p2mpkh          = unlockScriptProtocolP2MPKHGen.sample.get
+          val otherAssetInput = buildAssetInput(outputRef, p2mpkh)
+
+          val result = InputAddressUtil.addressFromProtocolInputs(otherAssetInput +: assetInputs)
+
+          result is None
+      }
+    }
+  }
+
+  trait Fixture {
+    def buildAssetInput(outputRef: api.model.OutputRef, unlockScript: protocol.vm.UnlockScript) = {
+      api.model.AssetInput(
+        outputRef,
+        serialize(unlockScript)
+      )
+    }
+  }
```

Comment:
Same here. This `Fixture` being stateless means we can pull out `buildAssetInput` to just being a plain function sitting somewhere in tests?

Reply:
done